### ISSUE TITLE
PyTorch NaN behavior and API design

### ISSUE THREAD

**Issue by heitorschueroff** [pytorch/pytorch#46544]
This issue is for discussing how PyTorch should handle NaN values and how we should design our operator API to do that. 

We have many issues such as #46225, reporting inconsistent handling of NaN values. One solution is to follow NumPy's design and have nan* ops which ignore NaN values while the regular versions treat NaN values as the largest element. However, this approach requires defining many new overloaded functions in native_functions which is not ideal. One alternative approach is to add an optional flag parameter `nan_policy` as is done in SciPy.

**Comment by mruberry** [comment]
We should implement the NumPy signatures but `nan_policy` is probably the more maintainable approach.

**Comment by mruberry** [comment]
To elaborate: my suggestion is to add the `nan_policy` kwarg to existing operators and either alias the nan* operators to call the operator with a particular nan_policy value, or have the nan_policy value determine which operator is called. We should be careful that whatever approach we take (and it may very operator-by-operator) doesn't increase dispatch overhead unless absolutely necessary.

**Comment by heitorschueroff** [comment]
Another reason we should add a nan_policy kwarg is for operators like `torch.mode` that do not have a NumPy equivalent and the SciPy equivalent does have nan_policy. See the following issue for a discussion on this https://github.com/pytorch/pytorch/issues/46225.

**Comment by heitorschueroff** [comment]
Proposed solution is to follow SciPy’s nan_policy (https://docs.scipy.org/doc/scipy/reference/dev/api-dev/nan_policy.html#f1), for reductions this means
* `nan_policy='raise'` should raise an error if the input has NaNs
* `nan_policy='propagate'` should propagate NaNs to the output. This is the in most of NumPy's reductions
* `nan_policy='omit'` should ignore all NaNs when computing the result
  * `reduce(a, nan_policy='omit')` should be equivalent to `reduce(a[~a.isnan()])`
  * this is the behavior of NumPy's `nan*` operators
* maybe we should add one other policy to treat NaNs as a valid value that is ordered as the largest value which can be useful for sorting and counting operations such as `median` and `mode`.

_Implementing this policy would change some reduction operators' default behavior to always propagate NaNs._

@mruberry @rgommers What do you think of this proposed solution? We should make sure that this is aligned with the Python Array API Standard for handling NaNs.

**Comment by mruberry** [comment]
This approach sounds good to me. It will be a BC-breaking change if we change the (currently implicit) nan policies of operators like mode, and we'll have to think carefully how impactful that deprecation is

**Comment by rgommers** [comment]
It sounds like a good idea to me too. At least that's a consistent design that can be referenced. It does raise the question about what set of operators we are talking about. In SciPy it's specific to the `stats` submodule. Are you proposing to do it for all reductions in PyTorch, or just on a case-by-case basis?

> maybe we should add one other policy to treat NaNs as a valid value that is ordered as the largest value which can be useful for sorting and counting operations such as `median` and `mode`.

Does that need a new mode? This is already what the default in NumPy is:
```
In [1]: x = np.array([2, 3, np.nan, 4, 5, np.nan])

In [2]: np.sort(x)
Out[2]: array([ 2.,  3.,  4.,  5., nan, nan])
```
It's not a reduction, so I'm not sure why it would need a `nan_policy`.

> @mruberry @rgommers What do you think of this proposed solution? We should make sure that this is aligned with the Python Array API Standard for handling NaNs.

There's no such thing in the array API standard. NaN's simply propagate there, just like `np.sum` and `torch.sum` (and following IEEE-754 whenever possible). We discussed the `nan*` functions early on, but decided that that was too specific and not a great design to begin with.

**Comment by heitorschueroff** [comment]
> It sounds like a good idea to me too. At least that's a consistent design that can be referenced. It does raise the question about what set of operators we are talking about. In SciPy it's specific to the `stats` submodule. Are you proposing to do it for all reductions in PyTorch, or just on a case-by-case basis?
> 

Probably for all reductions as I expect the code could be shared for the implementation.

> > maybe we should add one other policy to treat NaNs as a valid value that is ordered as the largest value which can be useful for sorting and counting operations such as `median` and `mode`.
> 
> Does that need a new mode? This is already what the default in NumPy is:
> 
> ```
> In [1]: x = np.array([2, 3, np.nan, 4, 5, np.nan])
> 
> In [2]: np.sort(x)
> Out[2]: array([ 2.,  3.,  4.,  5., nan, nan])
> ```
> 
> It's not a reduction, so I'm not sure why it would need a `nan_policy`.

While NumPy does sort `nan` as the largest value, it propagates `nan` for median and SciPy propagates `nan` for mode. If one wanted to know if `nan` is actually the mode, there would be no direct way to do it without this extra `nan_policy`. However, I'm not sure how often this use case rises so we can probably get away without adding an extra policy.

> > @mruberry @rgommers What do you think of this proposed solution? We should make sure that this is aligned with the Python Array API Standard for handling NaNs.
> 
> There's no such thing in the array API standard. NaN's simply propagate there, just like `np.sum` and `torch.sum` (and following IEEE-754 whenever possible). We discussed the `nan*` functions early on, but decided that that was too specific and not a great design to begin with.

I think there is a point here to be made against `nan*` operations because they can usually be easily achieved by applying a mask to the operation. A more general solution would be to implement masked tensors similar to [np.ma](https://numpy.org/doc/stable/reference/maskedarray.generic.html) allowing reductions to be performed while ignoring nan values.

**Comment by cpuhrsch** [comment]
As a drive by comment, since I don't see it mentioned: Similar to [deterministic mode](https://pytorch.org/docs/stable/notes/randomness.html) nan propagation policies could also be set as a top level flag or context manager like [inference_mode](https://pytorch.org/docs/stable/generated/torch.inference_mode.html). Not saying this is necessarily a great idea, but it's something we do.
